### PR TITLE
Updated the  --get-file method to get large files from NFS shares

### PR DESCRIPTION
**Pull request by ledrypotato**
The current implementation of the `get_file` method in `nfs.py` would only read up to 32768 bytes in size. I'm guessing this is because of how `nfs3.read` method is implemented.

To fix it this, I simply added a loop to fetch the entire contents of the file.

### Example of retrieving a 1.7MB image

Before the fix, only 32768 bytes are downloaded:

image

After the fix, the entire file is retrieved:

nxc-fix

### CODE REVIEW

File: nxc/protocols/nfs.py [hunk 1]
```diff
@@ -273,19 +273,38 @@ def get_file(self):
...
             file_handle = dir_data["resok"]["object"]["data"]
-            file_data = self.nfs3.read(file_handle, auth=self.auth)
 
-            if "resfail" in file_data:
-                raise Exception("Insufficient Permissions")
-            else:
-                data = file_data["resok"]["data"]
+            # Get the file attributes (size)
+            file_attrs = self.nfs3.getattr(file_handle, auth=self.auth)
+            file_size = file_attrs["attributes"]["size"]
```
Comment (NeffIsBack) [inline nxc/protocols/nfs.py:282]: You can get the file size from the lookup above. It should be in `dir_data["resok"]["obj_attributes"]["attributes"]["size"]`.

File: nxc/protocols/nfs.py [hunk 2]
```diff
@@ -273,19 +273,38 @@ def get_file(self):
...
+
+            # Handle files over 32768 bytes
+            offset = 0
+            chunk_size = 32768  # 32KB
+            eof = False
```
Comment (NeffIsBack) [inline nxc/protocols/nfs.py:287]: In the read response is also the variable `eof` present. Can you check that in the loop condition and handle it somehow if we get an EOF, but the size is still below the actual file_size?

File: nxc/protocols/nfs.py [hunk 3]
```diff
@@ -273,19 +273,38 @@ def get_file(self):
...
-                raise Exception("Insufficient Permissions")
-            else:
-                data = file_data["resok"]["data"]
+            # Get the file attributes (size)
+            file_attrs = self.nfs3.getattr(file_handle, auth=self.auth)
+            file_size = file_attrs["attributes"]["size"]
+
+            # Handle files over 32768 bytes
+            offset = 0
+            chunk_size = 32768  # 32KB
```
Comment (NeffIsBack) [inline nxc/protocols/nfs.py:286]: The default chunk size seems to be limited by `1024 * 1024` which is also the default value for the read function.
Can you check if you can for sure only download 32KB at a time? Do you have an idea why that could be? OS specific (which one are you running)?
image

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Great catch! Thanks for the bug report and the PR!

File: nxc/protocols/nfs.py (hunk 3 above)
Comment (ledrypotato) [reply nxc/protocols/nfs.py]: Thanks for getting back to me. I am doing my tests on Kali host, more specifically in an Exegol container in it.

You are right, I modified the code to handle the size with `file_size = dir_data["resok"]["obj_attributes"]["attributes"]["size"]`.

I also implemented the file read with the EOF that is returned from the `read` method.

This is my first PR, so excuse me if I don't know exactly how the process goes. Do I have to make another PR with the changes that I just made or can I add them on to the current one here?

Comment (termanix) [reply nxc/protocols/nfs.py]: > Thanks for getting back to me. I am doing my tests on Kali host, more specifically in an Exegol container in it.
> 
> You are right, I modified the code to handle the size with `file_size = dir_data["resok"]["obj_attributes"]["attributes"]["size"]`.
> 
> I also implemented the file read with the EOF that is returned from the `read` method.
> 
> This is my first PR, so excuse me if I don't know exactly how the process goes. Do I have to make another PR with the changes that I just made or can I add them on to the current one here?

You can add it on your this PR. Thank you btw

Comment (ledrypotato) [reply nxc/protocols/nfs.py]: Ok I think I added it correctly to this PR. Thanks to you for making the NFS protocol with nxc 😉

**Comment by termanix** [comment]
Default chunk_count aldready 1024*1024. No need to re-enter the default value. And also yes, we miss that part 😄

But of course Neff will have the final words 😄. But although the default chunk count is approximately 1 MB, I able to download a 2MB file...

image

**Comment by ledrypotato** [comment]
> Default chunk_count aldready 1024*1024. No need to re-enter the default value. And also yes, we miss that part 😄
> 
> But of course Neff will have the final words 😄. But although the default chunk count is approximately 1 MB, I able to download a 2MB file...
> 
> image

So you managed to download your 2MB file (ProcMon64.exe) without the changes I made? That's weird if so, I thought the file read was capped at 1024 * 1024 bytes... Maybe it's already handled it somewhere else in the code and I didn't see it 🤔

**Comment by termanix** [comment]
> > Default chunk_count aldready 1024*1024. No need to re-enter the default value. And also yes, we miss that part 😄
> > But of course Neff will have the final words 😄. But although the default chunk count is approximately 1 MB, I able to download a 2MB file...
> > image
> 
> So you managed to download your 2MB file (ProcMon64.exe) without the changes I made? That's weird if so, I thought the file read was capped at 1024 * 1024 bytes... Maybe it's already handled it somewhere else in the code and I didn't see it 🤔

No not :D I can download it full size with your changes. Look at my ss carefully, above the red line without your changes, below with your changes.

**Comment by ledrypotato** [comment]
> > > Default chunk_count aldready 1024*1024. No need to re-enter the default value. And also yes, we miss that part 😄
> > > But of course Neff will have the final words 😄. But although the default chunk count is approximately 1 MB, I able to download a 2MB file...
> > > image
> > 
> > 
> > So you managed to download your 2MB file (ProcMon64.exe) without the changes I made? That's weird if so, I thought the file read was capped at 1024 * 1024 bytes... Maybe it's already handled it somewhere else in the code and I didn't see it 🤔
> 
> No no 😄 I can download it full size with your changes. Look at my ss carefully, above the red line without your changes, below with your changes. My previous comment should have explained in more detail, my mistake.

Oh I understand, it's weird that you managed to download up to 524,288 bytes without the changes though where I could only download 32768 bytes... 

524,288 is half the value of 1024 * 1024. There must be something else handling how much of the file is downloaded in the original pyNfsClient.

**Comment by NeffIsBack** [comment]
> > No no 😄 I can download it full size with your changes. Look at my ss carefully, above the red line without your changes, below with your changes. My previous comment should have explained in more detail, my mistake.
> 
> Oh I understand, it's weird that you managed to download up to 524,288 bytes without the changes though where I could only download 32768 bytes...
> 
> 524,288 is half the value of 1024 * 1024. There must be something else handling how much of the file is downloaded in the original pyNfsClient.

Very weird, i was able to pull the full MB (1024*1024) in one command. But as long as we have proper handling of the possible states we should be good

**Comment by NeffIsBack** [comment]
But i agree with @termanix, there is no need in specifying the `chunk_count` again. It is already the default in the library. When this is fixed we should be good to go.

And yes, just push the changes here in the PR, thank you for your work! :)

**Comment by NeffIsBack** [comment]
@termanix mentioned to me that we could run into problems if we don't write the data immediately to the file. Especially with large files (thinking of server backups here) i can see that getting problematic.

@ledrypotato pushed the fix and some formatting changes, so we should be good to go now.

**Review by NeffIsBack — APPROVED** [review]
LGTM:
image

**Comment by ledrypotato** [comment]
Good point. Thanks for cleaning up the code! > @termanix mentioned to me that we could run into problems if we don't write the data immediately to the file. Especially with large files (thinking of server backups here) i can see that getting problematic.
> 
> @ledrypotato pushed the fix and some formatting changes, so we should be good to go now.

**Pull request merged**